File: evals/persistence.py
"""Durable eval run history (CV9.E2.S19 / AI-11).

Each ``eval <name>`` run is appended as one JSONL record under
``<mirror_home>/eval-history/<eval_name>.jsonl`` (or a gitignored repo-local
fallback when no mirror home is configured) so runs can be trended instead of
vanishing after stdout. Persistence is deliberately fail-soft: an eval's report
and exit code must never depend on whether the history write succeeded.
"""

from __future__ import annotations

import json
import logging
from dataclasses import asdict, dataclass, field
from pathlib import Path

from memory.config import resolve_mirror_home

logger = logging.getLogger(__name__)

# Repo-local fallback for a bare checkout with no configured mirror home.
# Eval history is non-deterministic run noise and must never be committed —
# see .gitignore.
_FALLBACK_DIR = Path(__file__).resolve().parent / ".history"


@dataclass
class EvalRunRecord:
    """One persisted eval run.

    ``model``/``prompt_hash`` are ``None`` for a genuinely prompt-free eval
    (e.g. keyword-based routing, pure scoring math) — a distinguishable state
    from "hash unchanged", not a fake/blank hash.
    """

    eval_name: str
    started_at: str
    ended_at: str
    model: str | None
    prompt_hash: str | None
    score: float
    threshold: float
    passed: bool
    probes: list[dict] = field(default_factory=list)
    schema_version: int = 1


def history_path(eval_name: str) -> Path:
    """Resolve the JSONL history file for one eval.

    Prefers ``<mirror_home>/eval-history/``; falls back to the gitignored
    repo-local ``evals/.history/`` when no mirror home is configured.
    """
    try:
        base = resolve_mirror_home() / "eval-history"
    except ValueError:
        base = _FALLBACK_DIR
    return base / f"{eval_name}.jsonl"


def append_run(record: EvalRunRecord) -> None:
    """Append one run to its eval's history file. Never raises.

    A persistence failure (unwritable path, disk issue) must never affect the
    eval's own report or exit code — the probe result is the point.
    """
    try:
        path = history_path(record.eval_name)
        path.parent.mkdir(parents=True, exist_ok=True)
        with path.open("a", encoding="utf-8") as f:
            f.write(json.dumps(asdict(record)) + "\n")
    except Exception:
        logger.warning("eval history append failed for %s", record.eval_name, exc_info=True)


def read_history(eval_name: str, limit: int = 10) -> list[EvalRunRecord]:
    """Return up to ``limit`` most-recent records, newest first.

    Tolerates a malformed trailing line (e.g. a crash mid-write) by skipping
    it rather than failing the whole read.
    """
    path = history_path(eval_name)
    if not path.exists():
        return []
    records: list[EvalRunRecord] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            records.append(EvalRunRecord(**json.loads(line)))
        except (json.JSONDecodeError, TypeError):
            continue
    return list(reversed(records))[:limit]

File: ts/evals/harness/persistence.ts
/**
 * Durable eval run history (CV22.DS10.TS3, port of `evals/persistence.py`).
 *
 * Each run is appended as one JSONL record under
 * `<mirror_home>/eval-history/<eval_name>.jsonl` -- the same directory, file
 * names, and field set the Python harness wrote, so a TypeScript run lands
 * beneath the Python runs already there and `--history` trends across the
 * cutover instead of starting a second log.
 *
 * Persistence is deliberately fail-soft: an eval's report and exit code must
 * never depend on whether the history write succeeded.
 *
 * Retention rule (CV22.DS10.TS3): **deleting a module never deletes its
 * measurements.** `scene.jsonl`, `routing.jsonl`, and `retrieval.jsonl` stay on
 * disk after their modules retire, and `--history <name>` still renders them,
 * because a history read opens a file rather than a module.
 */

import { appendFileSync, existsSync, mkdirSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { type DbPathEnv, resolveMirrorHome } from "#frontDoor/dbPath.ts";

/** Repo-local fallback for a checkout with no configured mirror home. */
const FALLBACK_DIR = join(dirname(dirname(fileURLToPath(import.meta.url))), ".history");

/** One persisted probe verdict inside a run record. */
export interface EvalProbeRecord {
  id: string;
  passed: boolean;
  notes: string;
  /** Present from schema 3; absent in Python-era records. */
  blocking?: boolean;
}

/**
 * One persisted eval run. Field names are snake_case because this is a wire
 * contract with the Python-era log, not an internal type.
 *
 * `model`/`prompt_hash` are null for a genuinely prompt-free eval -- a
 * distinguishable state from "hash unchanged", not a blank hash.
 *
 * `suite_run_id` correlates the per-eval records written by one `--all`
 * invocation; null for a standalone run.
 *
 * `schema_version` is **3** for records written under blocking semantics
 * (CV22.DS10.TS3). Versions 2 and 3 share a top-level field set and differ in
 * what `passed` means: at 2 it was `score >= threshold`; at 3 it also accounts
 * for failed blocking probes. Python's reader builds `EvalRunRecord(**line)`
 * from the same fields, so it keeps parsing v3 lines for as long as it exists.
 */
export interface EvalRunRecord {
  eval_name: string;
  started_at: string;
  ended_at: string;
  model: string | null;
  prompt_hash: string | null;
  score: number;
  threshold: number;
  passed: boolean;
  probes: EvalProbeRecord[];
  schema_version: number;
  suite_run_id: string | null;
}

/**
 * Resolve the JSONL history file for one eval.
 *
 * Prefers `<mirror_home>/eval-history/` through the same resolver the front
 * door uses (MIRROR_HOME wins; MIRROR_USER derives `~/.mirror-minds/<user>`
 * with the legacy `~/.mirror/<user>` fallback), so the harness cannot silently
 * open a second log beside the Python records. Falls back to the gitignored
 * repo-local `ts/evals/.history/` when no mirror home is configured.
 */
export function historyPath(evalName: string, env: DbPathEnv = process.env): string {
  let base: string;
  try {
    base = join(resolveMirrorHome(env), "eval-history");
  } catch {
    base = FALLBACK_DIR;
  }
  return join(base, `${evalName}.jsonl`);
}

/**
 * Append one run to its eval's history file. Never throws.
 *
 * A persistence failure (unwritable path, disk issue) must never affect the
 * eval's own report or exit code -- the probe result is the point.
 */
export function appendRun(record: EvalRunRecord, env: DbPathEnv = process.env): void {
  try {
    const path = historyPath(record.eval_name, env);
    mkdirSync(dirname(path), { recursive: true });
    appendFileSync(path, `${JSON.stringify(record)}\n`, "utf8");
  } catch {
    // Swallowed by contract. The report is the product of a run; the log is not.
  }
}

/**
 * Return up to `limit` most-recent records, newest first.
 *
 * Tolerates a malformed line (a crash mid-write) by skipping it rather than
 * failing the whole read, and accepts both schema 2 and schema 3 records.
 */
export function readHistory(
  evalName: string,
  limit = 10,
  env: DbPathEnv = process.env,
): EvalRunRecord[] {
  const path = historyPath(evalName, env);
  if (!existsSync(path)) return [];
  const records: EvalRunRecord[] = [];
  for (const rawLine of readFileSync(path, "utf8").split("\n")) {
    const line = rawLine.trim();
    if (!line) continue;
    try {
      records.push(JSON.parse(line) as EvalRunRecord);
    } catch {
      // Skip the bad line; the good records around it still trend.
    }
  }
  return records.reverse().slice(0, limit);
}
